### PR TITLE
Implement edition-based macro :pat feature

### PR DESCRIPTION
**Pull request by mark-i-m**
This PR does two things:
1. Fixes the perf regression from https://github.com/rust-lang/rust/pull/80100#issuecomment-750893149
2. Implements `:pat2018` and `:pat2021` matchers, as described by @joshtriplett  in https://github.com/rust-lang/rust/issues/54883#issuecomment-745509090 behind the feature gate `edition_macro_pat`.

r? @petrochenkov 

cc @Mark-Simulacrum

### CODE REVIEW

File: src/test/ui/macros/macro-follow.rs [hunk 1]
```diff
@@ -5,23 +5,23 @@
...
+    ($p:pat $q:pat) => {};   //~ERROR `$p:pat2018` is followed by `$q:pat2018`
+    ($p:pat $e:expr) => {};  //~ERROR `$p:pat2018` is followed by `$e:expr`
+    ($p:pat $t:ty) => {};    //~ERROR `$p:pat2018` is followed by `$t:ty`
+    ($p:pat $s:stmt) => {};  //~ERROR `$p:pat2018` is followed by `$s:stmt`
+    ($p:pat $q:path) => {};  //~ERROR `$p:pat2018` is followed by `$q:path`
+    ($p:pat $b:block) => {}; //~ERROR `$p:pat2018` is followed by `$b:block`
+    ($p:pat $i:ident) => {}; //~ERROR `$p:pat2018` is followed by `$i:ident`
+    ($p:pat $t:tt) => {};    //~ERROR `$p:pat2018` is followed by `$t:tt`
+    ($p:pat $i:item) => {};  //~ERROR `$p:pat2018` is followed by `$i:item`
+    ($p:pat $m:meta) => {};  //~ERROR `$p:pat2018` is followed by `$m:meta`
```
Comment (petrochenkov) [inline src/test/ui/macros/macro-follow.rs:24]: This looks like a regression in diagnostics to me.

File: compiler/rustc_parse/src/parser/nonterminal.rs [hunk 2]
```diff
@@ -141,12 +133,11 @@ impl<'a> Parser<'a> {
                     }
                 }
             }
-            NonterminalKind::Pat => {
-                let (mut pat, tokens) = self.collect_tokens(|this| match or_pat_mode {
-                    OrPatNonterminalMode::TopPat => {
-                        this.parse_top_pat(GateOr::Yes, RecoverComma::No)
-                    }
-                    OrPatNonterminalMode::NoTopAlt => this.parse_pat(None),
+            pat_kind @ (NonterminalKind::Pat2018 | NonterminalKind::Pat2021) => {
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/nonterminal.rs:136]:
```suggestion
            NonterminalKind::Pat2018 | NonterminalKind::Pat2021 => {
```
`kind` is already available.

File: compiler/rustc_parse/src/parser/nonterminal.rs [hunk 3]
```diff
@@ -62,7 +58,7 @@ impl<'a> Parser<'a> {
                 },
                 _ => false,
             },
-            NonterminalKind::Pat => match token.kind {
+            pat_kind @ (NonterminalKind::Pat2018 | NonterminalKind::Pat2021) => match token.kind {
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/nonterminal.rs:61]:
```suggestion
            NonterminalKind::Pat2018 | NonterminalKind::Pat2021 => match token.kind {
```
`kind` is already available.

File: compiler/rustc_ast/src/token.rs [hunk 4]
```diff
@@ -703,12 +704,18 @@ pub enum NonterminalKind {
 }
 
 impl NonterminalKind {
-    pub fn from_symbol(symbol: Symbol) -> Option<NonterminalKind> {
+    pub fn from_symbol(symbol: Symbol, span: Span) -> Option<NonterminalKind> {
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/token.rs:707]: Could you pass an edition closure like in #80272 instead of the whole span?
(+ a comment telling that `span.edition()` is relatively expensive.)

File: src/test/ui/macros/macro-follow.rs (hunk 1 above)
Comment (petrochenkov) [reply src/test/ui/macros/macro-follow.rs:24]: In general `pat2018` and `pat2021` look shameful enough to hide them as much as possible and not show to users unless really necessary.

Comment (mark-i-m) [reply src/test/ui/macros/macro-follow.rs:24]: Hmm... ok, I was wondering what you would think of it. I'll change it.

File: src/test/ui/macros/macro-input-future-proofing.rs [hunk 5]
```diff
@@ -8,7 +8,8 @@ macro_rules! errors_everywhere {
     ( { $ty:ty } ) => ();
     ( [ $ty:ty ] ) => ();
     ($bl:block < ) => ();
-    ($pa:pat >) => (); //~ ERROR `$pa:pat` is followed by `>`, which is not allowed for `pat`
+    ($pa:pat >) => ();
+    //~^ ERROR `$pa:pat` is followed by `>`, which is not allowed for `pat`
```
Comment (petrochenkov) [inline src/test/ui/macros/macro-input-future-proofing.rs:12]: Nit: unnecessary change.

**Comment by petrochenkov** [comment]
r=me with commits squashed.

File: src/test/ui/macros/macro-input-future-proofing.rs (hunk 5 above)
Comment (mark-i-m) [reply src/test/ui/macros/macro-input-future-proofing.rs]: Oops good catch. Fixed.

**Comment by mark-i-m** [comment]
@petrochenkov Fixed nit and squashed.
